cmd/config/crypto.go

```diff
@@ -0,0 +1,323 @@
+// the S3 secret key.
+type KeyDerivationParams struct {
+	Algorithm  string `json:"algorithm"`
+	Salt       []byte `json:"salt"`
+	TimeCost   int    `json:"time"`
+	MemoryCost int    `json:"memory"`
+	Threads    int    `json:"threads"`
+}
+
+// DefaultKeyDerivationParams returns new Argon2id key
+// derivation parameters with the given salt and reasonable
+// default cost parameters.
+func DefaultKeyDerivationParams(salt []byte) KeyDerivationParams {
+	return KeyDerivationParams{
+		Algorithm:  "Argon2id",
+		Salt:       salt,
+		TimeCost:   1,
+		MemoryCost: 64 * 1024,
+		Threads:    4,
+	}
+}
+
+// MasterKey is a 256 bit cryptographic key that can be used to
+// encrypt data streams.
+type MasterKey []byte
+
+// DeriveMasterKey derives a new master key from the given password
+// and key deriviation parameters.
+//
+// In the context of S3, the password should be the S3 secret key.
+//
+// It returns an error if the key derivation parameters are invalid
+// or would cause a very expensive key derivation.
+func DeriveMasterKey(password string, params KeyDerivationParams) (MasterKey, error) {
+	// Define some reasonable max. values. We should not
+	// accept arbitrary cost parameters (from untrusted sources)
+	// since large cost parameters can exhaust server resources.
+	const (
+		MaxTimeCost   = 10         // At most 10 passes over the memory
+		MaxMemoryCost = 256 * 1024 // At most 256 MiB memory
+		MaxThreads    = 16         // At most 16 threads
+	)
+
+	if params.Algorithm != "Argon2id" {
+		return nil, errors.New("config: invalid key derivation algorithm")
+	}
+	if params.TimeCost < 1 || params.TimeCost > MaxTimeCost {
+		return nil, errors.New("config: invalid key derivation time cost")
+	}
+	if params.MemoryCost > MaxMemoryCost {
+		return nil, errors.New("config: invalid key derivation memory cost")
+	}
+	if params.Threads < 1 || params.Threads > MaxThreads {
+		return nil, errors.New("config: invalid key derivation parallelism")
+	}
+	return MasterKey(argon2.IDKey(
+		[]byte(password),
+		params.Salt,
+		uint32(params.TimeCost),
+		uint32(params.MemoryCost),
+		uint8(params.Threads),
+		32,
+	)), nil
+}
+
+// EncryptBytes encrypts and authenticates the given data
+// and returns the encrypted ciphertext.
+func (key MasterKey) EncryptBytes(data []byte) ([]byte, error) {
+	ciphertext := bytes.NewBuffer(make([]byte, 0, len(data)))
+
+	w, err := key.Encrypt(ciphertext)
+	if err != nil {
+		return nil, err
+	}
+	if _, err = io.Copy(w, bytes.NewReader(data)); err != nil {
+		return nil, err
+	}
+	if err = w.Close(); err != nil {
+		return nil, err
+	}
+	return ciphertext.Bytes(), nil
+}
+
+// Encrypt returns a new io.WriterCloser that wraps w and
+// encrypts and authenticates everything written to it.
+//
+// The returned io.WriteCloser has to be closed successfully to
+// complete the encryption process.
+func (key MasterKey) Encrypt(w io.Writer) (io.WriteCloser, error) {
+	iv, err := sioutil.Random(16)
```

Comment:
Check if 'key' has a useful value. Maybe sanity check against all zeros.

Reply:
> Check if 'key' has a useful value. Maybe sanity check against all zeros.

This is not really possible. Also,I get the point that a check against the default value would eventually catch a un-handled error higher up the stack. The caveat is that the default value is `nil`. All of these values (`all 0's`, `nil`, `all 1's`, ...) are equally bad. 

The value of the `MasterKey` type  - as any other crypto. key -  has to be chosen at random.
It's not the responsibility of this function to check that the `MasterKey` is chosen properly - simply b/c it cannot in any sufficient manner.
For the same reason `hmac.New(...)` does not fail if you provide a `nil` key or an all `1's` key.